### PR TITLE
Feature/boardkeys #2449

### PR DESCRIPTION
**Pull request by road42**
# Implementation of board keys

## Every board gets a key (autogenerated during migration from A-Z)
The existing view settings are kept. Nothing changes at first.

## When you create a board you have to enter a key
1-createBoard

## Filled board creation
2-createBoardFilled
The length is validated

## The created board
3-created
At first nothing changed

## The boardAdmin may change how the board is shown
4-changeSettings

## There are different settings
5-setBoardKey
You may request more :)

## The boardTitle with the new settings
6-exampleView

## The boardList with the new settings
7-boardListView

---
This change is [Reviewable](https://reviewable.io/reviews/wekan/wekan/2526)

### CODE REVIEW

**Comment by justinr1234** [comment]
@road42 looks fantastic!

File: models/boards.js [hunk 1]
```diff
@@ -10,6 +10,66 @@ Boards.attachSchema(new SimpleSchema({
      */
     type: String,
   },
+  boardKey: {
+    /**
+     * The key of the board
+     */
+    type: String,
+    min: 1,
+    max: 3,
```
Comment (justinr1234) [inline models/boards.js]: This should be larger than 3 maximum. It should be a configurable setting in the admin section. The `max` key can take a function that returns what the maximum is. You can use this to lookup the admin setting for maximum. Make the admin maximum setting readable by all users but only writeable by admins.

File: models/boards.js [hunk 2]
```diff
@@ -10,6 +10,66 @@ Boards.attachSchema(new SimpleSchema({
...
+    autoValue() { // eslint-disable-line consistent-return
+
+      // Fetch required fields
+      const title = this.field('title');
+      const boardKey = this.field('boardKey');
+      const showBoardKey = this.field('showBoardKey');
+
+      // Only act on insert or update
+      // and then all three required fields are set
+      if ((this.isInsert || this.isUpdate) && !this.isSet && title.isSet && boardKey.isSet && showBoardKey.isSet) {
```
Comment (justinr1234) [inline models/boards.js]: This "full title" should return a default value, not only returning a value when all the others are set. It should return a value for `this.field('title')` at the very least.

File: models/boards.js [hunk 3]
```diff
@@ -334,6 +394,18 @@ Boards.helpers({
       swimlane.copy(_id);
     });
   },
+  /**
+   * Generates random chars as board key
+   */
+  generateKey() {
+    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
```
Comment (justinr1234) [inline models/boards.js]: This function should check for the existence of the key. It should unique with regards to the rest of the keys in the system. With only 3 digits here, you're going to run out of unique keys very quickly. Additionally, this should default to the first 3 characters of the `title` being created for the board.

Example:

You have a board called `Test`

You create a board called `Test`

The first board should have a board key `TES` and the second one should have a board key `TEST`

You create another board called `Test`. It should have a board key called `TEST2`. Basically, you want to make sure the keys are unique and that they (hopefully) follow the name of the board as closely as possible.

**Comment by xet7** [comment]
@road42 

You can make new PR to master branch, that now has meteor 1.8.

I deleted old branches, so now there is only:
- master, that has new meteor 1.8.x
- devel, that has old meteor 1.6.x, because of snap. Sometime when snap is fixed, devel will also be removed, and development happens only at master branch.

**Pull request closed without merging**